File: cpp/datastruct/linkedlist.cpp
#include <stdio.h>

struct Node
{
  Node* next;
  int value;
  
  Node(int value) {
    this->value = value;
  }
};

struct LinkedList
{
  Node* head;

  LinkedList() {}

  LinkedList(Node* node) 
  {
    this->head = node;
  }

  ~LinkedList() 
  {
    if (!this->head) {
      return;
    }

    Node* tmp = this->head;
    Node* previousNode = this->head;
    while(tmp) {
      previousNode = tmp;
      tmp = tmp->next;
      printf("delete %d\n", previousNode->value);
      delete previousNode;
    }
  }

  void addNode(Node* node) {
    if (!this->head) {
      this->head = node;
      return;
    }

    Node* tmp = this->head;
    if (node->value < tmp->value) {
      node->next = tmp;
      this->head = node;
    } else {
      while(tmp->next && tmp->next->value < node->value) {
        tmp = tmp->next;
      }

      if (tmp->next)
      {
        node->next = tmp->next;
      }
      tmp->next = node;
    }
  }

  void deleteNode(Node* node) {
    if (!this->head) {
      return;
    }

    Node* tmp = this->head;
    if (node->value == tmp->value) {
      this->head = this->head->next;
    } else {
      while(tmp->next) {
        if (tmp->next->value == node->value)
        {
          Node* foundNode = tmp->next;
          tmp->next = tmp->next->next;
          delete foundNode;
          return;
        }
        tmp = tmp->next;
      }
    }
  }

  void print() {
    Node* tmp = this->head;
    printf("Print LinkedList\n");
    while(tmp) {
      printf("%d\n", tmp->value);
      tmp = tmp->next;
    }
    printf("\n");
  }
};

int main(int argc, char const *argv[])
{
  LinkedList* list = new LinkedList();
  Node* node1 = new Node(4);
  Node* node2 = new Node(5);
  Node* node3 = new Node(1);
  Node* node4 = new Node(26);
  Node* node5 = new Node(15);
  Node* node6 = new Node(65);
  Node* node7 = new Node(8);
  Node* node8 = new Node(102);

  list->addNode(node1);
  list->addNode(node2);
  list->addNode(node3);
  list->addNode(node4);
  list->addNode(node5);
  list->addNode(node6);
  list->addNode(node7);
  list->addNode(node8);

  list->print();

  list->deleteNode(node3);
  list->deleteNode(node6);

  list->print();

  delete list;
  return 0;
}

File: graph/linkedlist.py
class Node:
    data = None
    next = None

    def __init__(self, data):
        self.data = data

    def append(self, node):
        if self.next is None:
            self.next = node
        else:
            tmp = self
            while tmp.next is not None:
                tmp = tmp.next
            tmp.next = node
        return self


def delete(head, value):
    if head is None:
        print('Nothing to delete')
        return head

    if head.data == value:
        return head.next

    tmp = head
    while tmp.next is not None:
        if tmp.next.data == value:
            tmp.next = tmp.next.next
            return head
        tmp = tmp.next
    return head


def remove_duplicate(head):
    tmp = head
    dict = {head.data : 0}

    while tmp.next is not None:
        if tmp.next.data in dict:
            tmp.next = tmp.next.next
        else:
            dict[tmp.next.data] = 0
            tmp = tmp.next


def print_list(head):
    tmp = head
    print(tmp.data, end=' ')
    while tmp.next is not None:
        print(tmp.next.data, end=' ')
        tmp = tmp.next
    print()


def get_n_last(head, n):
    first_pointer = head
    second_pointer = None
    counter = 1

    while first_pointer.next is not None:
        counter += 1
        first_pointer = first_pointer.next
        if counter == n:
            second_pointer = head
        elif counter > n:
            second_pointer = second_pointer.next
    return second_pointer


def delete_node(node):
    if node.next is None:
        print("Can't to that")
        return

    node.data = node.next.data
    node.next = node.next.next


def add_list(list1, list2):
    first = list1
    second = list2
    zero = Node(0)

    value = (first.data + second.data) % 10
    tmp = int((first.data + second.data) / 10)

    head = Node(value)
    while first.next is not None or second.next is not None or tmp > 0:
        first = first.next if first.next is not None else zero
        second = second.next if second.next is not None else zero

        value = (first.data + second.data + tmp) % 10
        tmp = int((first.data + second.data + tmp) / 10)
        head.append(Node(value))
    return head


def add_list1(list1, list2):
    first = list1
    first_string = ''
    while first is not None:
        first_string = str(first.data) + first_string
        first = first.next

    second = list2
    second_string = ''
    while second is not None:
        second_string = str(second.data) + second_string
        second = second.next

    total = int(first_string) + int(second_string)
    head = None
    for char in str(total):
        if head is None:
            head = Node(int(char))
        else:
            head.append(Node(int(char)))
    return head


def detect_circular(head):
    slow = head
    fast = head.next

    while True:
        if fast is None:
            return None
        if fast == slow:
            return fast

        slow = slow.next
        fast = fast.next.next

# node1 = Node(1)
# node2 = Node(2)
# node3 = Node(3)
# node4 = Node(2)
# node5 = Node(1)
# node6 = Node(4)
# node7 = Node(5)
# node8 = Node(1)
# node9 = Node(1)
# node10 = Node(1)
# node1.append(node2).append(node3).append(node4).append(node5).append(node6).append(node7).append(node8).append(node9).append(node10)


# print_list(node1)
# delete(node1, 4)
# delete(node1, 3)
# remove_duplicate(node1)
# print_list(node1)
# delete_node(node7)
# print_list(node1)
# print(get_n_last(node1, 5).data)

# list1 = Node(3).append(Node(1))
# list2 = Node(5).append(Node(9)).append(Node(2))
# head = add_list1(list1, list2)
# print_list(head)

head = Node(0)
tmp = head
for i in range(1, 20):
    node = Node(i)
    tmp.next = node
    tmp = node

node = get_n_last(head, 10)
node.next = get_n_last(head, 13)
print(node.data)
print(detect_circular(head).data)


# print(detect_circular(nodeA).data)